File: backend-v2/app/routers/hitl.py
import json
import uuid
from datetime import datetime, timezone
from typing import Optional, List

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from ..auth import require_user
from ..database import get_session_factory, ensure_schema
from ..models import Doc, HitlReview, Base
from ..schemas import DocResponse, HitlReviewResponse, _extract_outline

router = APIRouter(prefix="/hitl", tags=["hitl"])


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _get_db(user: dict):
    user_id = user["sub"]
    ensure_schema(user_id, Base)
    return get_session_factory(user_id)()


def _to_response(r: HitlReview, doc_name: str, doc=None, include_current: bool = False) -> HitlReviewResponse:
    return HitlReviewResponse(
        id=r.id,
        doc_id=r.doc_id,
        doc_name=doc_name,
        proposed_payload=json.loads(r.proposed_payload),
        agent_pat_prefix=r.agent_pat_prefix,
        outcome=r.outcome,
        human_notes=r.human_notes,
        created_at=r.created_at,
        resolved_at=r.resolved_at,
        doc_current=DocResponse.from_doc(doc) if (include_current and doc) else None,
    )


@router.get("/reviews", response_model=List[HitlReviewResponse])
def list_reviews(
    outcome:      Optional[str] = Query(None, description="Omit=pending only, 'all'=all, or specific value"),
    doc_id:       Optional[str] = Query(None, description="Filter by doc ID"),
    submitted_by: Optional[str] = Query(None, description="Filter by agent_pat_prefix"),
    user: dict = Depends(require_user),
):
    db = _get_db(user)
    try:
        q = db.query(HitlReview)
        if outcome is None:
            q = q.filter(HitlReview.outcome == None)  # noqa: E711
        elif outcome != "all":
            q = q.filter(HitlReview.outcome == outcome)
        if doc_id:
            q = q.filter(HitlReview.doc_id == doc_id)
        if submitted_by:
            q = q.filter(HitlReview.agent_pat_prefix == submitted_by)
        reviews = q.order_by(HitlReview.created_at.desc()).all()
        result = []
        for r in reviews:
            doc = db.query(Doc).filter(Doc.id == r.doc_id).first()
            result.append(_to_response(r, doc.name if doc else "Deleted doc"))
        return result
    finally:
        db.close()


@router.get("/reviews/{review_id}", response_model=HitlReviewResponse)
def get_review(review_id: str, user: dict = Depends(require_user)):
    db = _get_db(user)
    try:
        r = db.query(HitlReview).filter(HitlReview.id == review_id).first()
        if not r:
            raise HTTPException(404, "Review not found")
        doc = db.query(Doc).filter(Doc.id == r.doc_id).first()
        return _to_response(r, doc.name if doc else "Deleted doc", doc=doc, include_current=True)
    finally:
        db.close()


class ResolveRequest(BaseModel):
    outcome:     str            # "approved" | "rejected" | "cancelled"
    human_notes: Optional[str] = None


@router.post("/reviews/{review_id}/resolve")
def resolve_review(review_id: str, body: ResolveRequest, user: dict = Depends(require_user)):
    # Untrusted PAT callers cannot resolve reviews - only browser users and trusted tokens
    if user["auth_method"] == "pat" and not user.get("pat_trusted"):
        raise HTTPException(403, "Only trusted tokens or browser users can resolve reviews")

    if body.outcome not in ("approved", "rejected", "cancelled"):
        raise HTTPException(400, "outcome must be one of: approved, rejected, cancelled")

    db = _get_db(user)
    try:
        r = db.query(HitlReview).filter(HitlReview.id == review_id).first()
        if not r:
            raise HTTPException(404, "Review not found")
        if r.outcome is not None:
            raise HTTPException(409, "Review already resolved")

        doc = db.query(Doc).filter(Doc.id == r.doc_id).first()
        if doc:
            if body.outcome == "approved":
                payload = json.loads(r.proposed_payload)
                for field, value in payload.items():
                    setattr(doc, field, value)
                if "body" in payload:
                    doc.note_outline = _extract_outline(doc.body or "")
            doc.hitl_status = None
            doc.updated_at = _now()

        r.outcome = body.outcome
        r.human_notes = body.human_notes
        r.resolved_at = _now()
        db.commit()
        return {"ok": True, "outcome": body.outcome}
    finally:
        db.close()

File: v3/backend/src/routes/hitl.rs
use std::sync::Arc;

use axum::{extract::{Path, Query, State}, http::StatusCode, Json};
use serde::Deserialize;

use crate::meta_db::{self, HitlReview};
use crate::models::{AuthMethod, AuthUser, DocResponse};
use crate::store::{file, AppState};

type Res<T> = Result<Json<T>, (StatusCode, Json<serde_json::Value>)>;
fn err(code: StatusCode, msg: &str) -> (StatusCode, Json<serde_json::Value>) {
    (code, Json(serde_json::json!({"detail": msg})))
}

#[derive(Deserialize, Default)]
pub struct ReviewsQuery {
    pub outcome: Option<String>,
    pub doc_id: Option<String>,
    pub submitted_by: Option<String>,
}

pub async fn list_reviews(
    State(state): State<Arc<AppState>>,
    user: AuthUser,
    Query(q): Query<ReviewsQuery>,
) -> Res<Vec<HitlReview>> {
    let user_root = state.user_root_dir(&user.sub);
    let pool = meta_db::init_user_meta_db(&user_root).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

    let all_outcomes = q.outcome.as_deref() == Some("all");
    let mut rows = meta_db::fetch_reviews(&pool).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

    if !all_outcomes {
        rows.retain(|r| r.outcome.is_none());
    }
    if let Some(doc_id) = &q.doc_id {
        rows.retain(|r| &r.doc_id == doc_id);
    }
    if let Some(submitted_by) = &q.submitted_by {
        rows.retain(|r| r.agent_pat_prefix.as_deref() == Some(submitted_by.as_str()));
    }

    Ok(Json(rows))
}

pub async fn get_review(
    State(state): State<Arc<AppState>>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Res<serde_json::Value> {
    let user_root = state.user_root_dir(&user.sub);
    let pool = meta_db::init_user_meta_db(&user_root).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

    let review = meta_db::fetch_review_by_id(&pool, &id).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?
        .ok_or_else(|| err(StatusCode::NOT_FOUND, "review not found"))?;

    let doc_id: uuid::Uuid = review.doc_id.parse()
        .map_err(|_| err(StatusCode::INTERNAL_SERVER_ERROR, "invalid doc_id in review"))?;
    let docs_dir = state.user_docs_dir(&user.sub);
    let doc_current = file::find_doc_path(&docs_dir, doc_id)
        .and_then(|path| file::parse_doc(&path).ok())
        .map(|d| serde_json::to_value(DocResponse::from(&d)).unwrap_or_default());

    let mut val = serde_json::to_value(&review).unwrap_or_default();
    if let serde_json::Value::Object(ref mut map) = val {
        map.insert("doc_current".to_string(), doc_current.unwrap_or(serde_json::Value::Null));
    }
    Ok(Json(val))
}

#[derive(Deserialize)]
pub struct ResolveRequest {
    pub outcome: String,
    pub human_notes: Option<String>,
}

pub async fn resolve_review(
    State(state): State<Arc<AppState>>,
    user: AuthUser,
    Path(id): Path<String>,
    Json(body): Json<ResolveRequest>,
) -> Result<StatusCode, (StatusCode, Json<serde_json::Value>)> {
    if user.auth_method == AuthMethod::Pat && !user.pat_trusted {
        return Err(err(StatusCode::FORBIDDEN, "only trusted tokens or browser users can resolve reviews"));
    }

    let user_root = state.user_root_dir(&user.sub);
    let pool = meta_db::init_user_meta_db(&user_root).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

    let review = meta_db::fetch_review_by_id(&pool, &id).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?
        .ok_or_else(|| err(StatusCode::NOT_FOUND, "review not found"))?;

    if review.outcome.is_some() {
        return Err(err(StatusCode::CONFLICT, "review already resolved"));
    }

    meta_db::resolve_hitl_review(&pool, &id, &body.outcome, body.human_notes.as_deref()).await
        .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

    let doc_id: uuid::Uuid = review.doc_id.parse()
        .map_err(|_| err(StatusCode::INTERNAL_SERVER_ERROR, "invalid doc_id"))?;
    let docs_dir = state.user_docs_dir(&user.sub);
    let file_name = state.index.get_file_name(&user.sub, doc_id)
        .or_else(|| file::find_doc_path(&docs_dir, doc_id)
            .and_then(|p| p.file_name().and_then(|f| f.to_str()).map(String::from)));

    if let Some(file_name) = file_name {
        let path = docs_dir.join(&file_name);
        if body.outcome == "approved" {
            let payload: crate::models::UpdateDocRequest = serde_json::from_str(&review.proposed_payload)
                .map_err(|e| err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;
            if let Ok(mut doc) = file::parse_doc(&path) {
                if let Some(b) = payload.body {
                    doc.body = b;
                    doc.note_outline = Some(file::compute_outline(&doc.body));
                }
                if let Some(s) = payload.status { doc.status = s.parse().unwrap_or_default(); }
                if let Some(p) = payload.priority { doc.priority = p.parse().unwrap_or_default(); }
                doc.hitl_status = None;
                doc.updated_at = chrono::Utc::now();
                file::write_doc(&docs_dir, &doc, Some(&file_name)).ok();
                crate::store::git::commit_file(&user_root, &format!("docs/{}", file_name), &format!("hitl approved: {}", doc.title)).ok();
                state.index.upsert(&user.sub, &doc, &file_name);
            }
        } else if let Ok(mut doc) = file::parse_doc(&path) {
            doc.hitl_status = None;
            doc.updated_at = chrono::Utc::now();
            file::write_doc(&docs_dir, &doc, Some(&file_name)).ok();
            state.index.upsert(&user.sub, &doc, &file_name);
        }
    }

    Ok(StatusCode::OK)
}
